### DISCUSSION TITLE
How to use PostMessage and BindObjectAsync in the same function

### DISCUSSION THREAD

**Discussion by phuocminh6464** [cefsharp/CefSharp#4404] [General]
I have a button on web and when clicking it will send data to my WPF app. What I want is after processing with that data, WPF will send result to web. Currently, I use **CefSharp.PostMessage** to push message from web to WPF and handler **JavascriptMessageReceived** event to received message. Then I use that data to calculate and register **JavascriptObjectRepository** to send result to web after process ending. But the app always get error  **_Uncaught (in promise) ReferenceError: bindingObj is not defined_** in JS at ` var result= bindingObj.getResult();`  when I click button in web. Anyone has solution? Thanks
Note: I use Cefsharp ver 109.1.100 and net7

```
var isBound = web.JavascriptObjectRepository.IsBound("bindingObj");
if (isBound)
{
     web.JavascriptObjectRepository.UnRegister("bindingObj");
}
web.JavascriptObjectRepository.Register("bindingObj", new JavaScriptInteractionObj(), options: BindingOptions.DefaultBinder);
```

```
async function doAction() {
            CefSharp.PostMessage({ "Type": "key", "Value": "value"});

            CefSharp.DeleteBoundObject("bindingObj");
            CefSharp.RemoveObjectFromCache("bindingObj");
            await CefSharp.BindObjectAsync("bindingObj");

            var result= bindingObj.getResult();
            result.then((value) => {
                console.log(value);
            });
        }
```

**Comment by amaitland** [comment]
> Then I use that data to calculate and register **JavascriptObjectRepository** to send result to web after process ending

Registering an object is expensive and is not really meant to be used in a dynamic register/unregister fashion. There's caching in the render process for performance reasons. I'd suggest using [ResolveObject](https://cefsharp.github.io/api/109.1.x/html/E_CefSharp_IJavascriptObjectRepository_ResolveObject.htm) event to instead. Just register your object once.

Without a complete example it's impossible to know what's going on.

> How to use PostMessage and BindObjectAsync in the same function

Whilst you can, you typically wouldn't mix and match.

For `PostMessage` you typically would pass a javascript function in as a param, you'll get a `IJavascriptCallback` in your `c#` code which you can use to async execute the `JS function`.

There's a basic example at https://github.com/cefsharp/CefSharp/issues/2775#issuecomment-498454221 and a unit tests at https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Test/PostMessage/PostMessageTests.cs#L39

**Comment by phuocminh6464** [comment]
Hi @amaitland, thanks for your reply but I have more concern

**What I want is:** Using `PostMessage` in js function and receiveing message at `JavascriptMessageReceived` event in C# code. Then check `if (type == "CheckPrinter")`, WPF app will async execute the `responseFunc` JS function to post message for printing. When receiving message with `"Type": "Print"`, WPF will call print function and send result to web after printing. I want web can automatic get that result immediately after printing finished.
**JS FUNCTION**
```
let responseFunc = function () {
                CefSharp.PostMessage({
                    "Type": "Print", Value: {
                        StampList: [{ "ID": "123", "Name": "Name1", "Status": "Not yet" },
                        { "ID": "456", "Name": "Name 2", "Status": "Not yet" }]
                    }
                });
            };
CefSharp.PostMessage({ "Type": "CheckPrinter", "Value": false, "Callback": responseFunc });
```
**C# CODE**
```
private async void OnBrowserJavascriptMessageReceived(object sender, JavascriptMessageReceivedEventArgs e)
        {
            var msg = e.ConvertMessageTo<PostMessageCallBack>();
            var callback = msg.Callback;
            var type = msg.Type;
            var value = msg.Value;

            if (type == "CheckPrinter")
            {              
                  // Do check printer connection
                  // If Printer connection is good
                  if (isConnected)
                  {
                       //Call a javascript function
                       var response = await callback.ExecuteAsync();
                       var success = response.Success; // true
                       var result = response.Result; // null
                       var mess = response.Message; // null

                       // Here can not get new message!!!
                       dynamic msgCallBack = e.ConvertMessageTo<PostMessagePrinting>(); 
                       var typeCallBack = msgCallBack.Type; //"CheckPrinter"
                       var valueCallBack = msgCallBack.Value; // null
                  }
            }
        }                    
```
**What is problem:** Now, I'm using the above code and can not get new message when run  `dynamic msgCallBack = e.ConvertMessageTo<PostMessagePrinting>();` . Value of variable `msgCallBack` is not change, it is same as variable `msg`. 
So what I need to do to get new message and send result to web to get real time?

**Reply by amaitland** [reply]
- Every `PostMessage` call raises a new instance of the `JavascriptMessageReceived` event
- Your `javascript callback` function needs to return a value if you want to populate the `JavascriptResponse.Result` property.

**Reply by phuocminh6464** [reply]
@amaitland Thanks you, I did it.
